design/operater-gateway-support-migration-design.md

```diff
@@ -0,0 +1,88 @@
+# Contour Operator Gateway API Support Migration
+
+Status: Draft
+
+## Abstract
+Currently the Contour Operator and Contour Controller have both started implementing Gateway API support.
+Both components contain controllers that reconcile GatewayClass and Gateway resources and that conflict when reconciling status.
+This document describes options for how we should proceed to untangle these.
+
+## Background
+Currently the state of the world is as below:
+
+### Contour Operator
+- Configured with a GatewayClass object reference (name) to reconcile
+- Includes GatewayClass controller
+  - Checks if named GatewayClass has the correct (hardcoded to `projectcontour.io/contour-operator`) controller string
+  - Checks for Contour CRD in the parameters ref of the GatewayClass, uses parameters for provisioning Contour instances
+  - Sets admitted status on object if all valid
+- Includes Gateway controller
+  - Checks if Gateway refers to correct GatewayClass and that GatewayClass has been admitted
+  - Sets up finalizer for Gateway
+  - Creates Contour deployment resources with the Gateway being reconciled named in Contour's ConfigMap
+  - Sets up finalizer for GatewayClass
+  - Sets up finalizer for Contour CRD referenced in GatewayClass
+  - Sets admitted status on object if all valid
+
+### Contour Controller
+- Needs to be configured with a GatewayClass controller name to ensure it reconciles the correct GatewayClass
+- Also needs a Gateway namespace/name which is being deprecated
+- Includes GatewayClass controller
+  - Expects parameter ref field to be unset, otherwise GatewayClass is not admitted
+  - Admits GatewayClass that has the configured controller name
+- Includes Gateway controller
+  - Checks if Gateway references Contour owned GatewayClass
+  - Passes off reconciliation to dag package Gateway API processor
+  - Gateway is validated, status set on it and relevant *Routes
+
+### Current Problems
+- Contour requires Gateway namespace/name *and* controller name, the latter of which the operator does not provide
+- Contour expects GatewayClass parameters ref to be unset while Operator expects it to reference a Contour CRD
```

Comment:
This touches on the user experience of who creates Contour.  Looking at the original design doc, it states that if parametersRef field is left unspecified, the Operator should create an instance of Contour with default settings.  Is that still the case with the current design?

Reply:
The Operator has not yet implemented that feature and Contour itself actually requires the parametersRef to be nil

Miscommunication and mixup between the two codebases